trino/trino.go

```diff
@@ -932,48 +971,185 @@ func (qr *driverRows) fetch(allowEOF bool) error {
 		}
 	}
 	if qr.columns == nil && len(qresp.Columns) > 0 {
+		for i := range qresp.Columns {
+			err = unmarshalArguments(&(qresp.Columns[i].TypeSignature))
+			if err != nil {
+				return fmt.Errorf("error decoding column type signature: %w", err)
+			}
+		}
 		qr.initColumns(&qresp)
 	}
 	qr.rowsAffected = qresp.UpdateCount
 	return nil
 }
 
+func unmarshalArguments(signature *typeSignature) error {
+	for i, argument := range signature.Arguments {
+		var payload interface{}
+		switch argument.Kind {
+		case KIND_TYPE:
+			payload = &(signature.Arguments[i].typeSignature)
+		case KIND_NAMED_TYPE:
+			payload = &(signature.Arguments[i].namedTypeSignature)
+		case KIND_LONG:
+			payload = &(signature.Arguments[i].long)
+		}
+		err := json.Unmarshal(argument.Value, payload)
+		if err != nil {
+			return err
+		}
+		switch argument.Kind {
+		case KIND_TYPE:
+			err = unmarshalArguments(&(signature.Arguments[i].typeSignature))
+		case KIND_NAMED_TYPE:
+			err = unmarshalArguments(&(signature.Arguments[i].namedTypeSignature.TypeSignature))
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 func (qr *driverRows) initColumns(qresp *queryResponse) {
 	qr.columns = make([]string, len(qresp.Columns))
 	qr.coltype = make([]*typeConverter, len(qresp.Columns))
 	for i, col := range qresp.Columns {
 		qr.columns[i] = col.Name
-		qr.coltype[i] = newTypeConverter(col.Type)
+		qr.coltype[i] = newTypeConverter(col.Type, col.TypeSignature)
 	}
 }
 
 type typeConverter struct {
-	typeName   string
-	parsedType []string // e.g. array, array, varchar, for [][]string
+	typeName      string
+	parsedType    []string
+	scanType      reflect.Type
+	typeSignature typeSignature
+	hasPrecision  bool
+	precision     int64
+	scale         int64
+	hasSize       bool
+	size          int64
+}
+
+func newTypeConverter(typeName string, signature typeSignature) *typeConverter {
+	result := &typeConverter{
+		typeName:      typeName,
+		parsedType:    getNestedTypes([]string{}, signature),
+		typeSignature: signature,
+	}
+	result.scanType = getScanType(result.parsedType)
+	switch signature.RawType {
+	case "char", "varchar":
+		if len(signature.Arguments) > 0 && signature.Arguments[0].Kind == "LONG" {
+			result.hasSize = true
+			result.size = signature.Arguments[0].long
+		}
+	case "decimal":
+		if len(signature.Arguments) > 1 {
+			result.hasPrecision = true
+			if signature.Arguments[0].Kind == "LONG" {
+				result.precision = signature.Arguments[0].long
+			}
+			if signature.Arguments[1].Kind == "LONG" {
+				result.scale = signature.Arguments[1].long
+			}
+		}
+	}
+	return result
 }
 
-func newTypeConverter(typeName string) *typeConverter {
-	return &typeConverter{
-		typeName:   typeName,
-		parsedType: parseType(typeName),
+func getNestedTypes(types []string, signature typeSignature) []string {
+	types = append(types, signature.RawType)
+	if len(signature.Arguments) > 0 {
```

Comment:
what about the case when there are more than single type argument?
You just look at the first one. It feels to me that the `if` should be `len(signature.Arguments) == 0`, or we should explicitly support only specific types here.

Reply:
This is an incremental improvement only to get the types from the response instead of parsing the text type and splitting on `(`. Can we address this in a follow-up PR?

Reply:
BTW there would be multiple arguments only for composite types like row and we wouldn't return scan types for them anyway, leaving scanning completely to the user.

Reply:
Ah, I read this again and you're right, it shouldn't only grab the first arg when there are multiple ones. I changed this to `== 1`.